### PR TITLE
Add option to build either shared or static libraries via CMake (update to CRTMv3 release/REL-3.1.1 branch)

### PR DESCRIPTION
**Pull request by BenjaminTJohnson**
update to CRTMv3 `release/REL-3.1.1` branch.       

- Introduced `BUILD_SHARED_LIBS` option to allow users to select between building shared or static libraries.
- Modified `CMakeLists.txt` to conditionally set the library type based on the value of `BUILD_SHARED_LIBS`.
- Simplified library creation by using the `LIBRARY_TYPE` variable, removing the need to always build both shared and static libraries.
- Updated target linking and installation settings to apply only to the selected library type.

Default is `BUILD_SHARED_LIBS=ON`. 

`cmake -DBUILD_SHARED_LIBS=ON ..`  produces <build>/src/libcrtm.so
`cmake -DBUILD_SHARED_LIBS=OFF ..`  produces <build>/src/libcrtm.a

This will likely break things, please test compiling/linking CRTM in applications.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -215,23 +215,31 @@ if(((CMAKE_BUILD_TYPE STREQUAL "Debug") OR (CMAKE_BUILD_TYPE STREQUAL "RelWithDe
...
 
+# Option to select between shared and static libraries
+option(BUILD_SHARED_LIBS "Build shared libraries instead of static ones" ON)
 
-add_library(${PROJECT_NAME} SHARED ${${PROJECT_NAME}_src_files})
-add_library(${PROJECT_NAME}_static STATIC ${${PROJECT_NAME}_src_files})
+# Determine the type of library to build
+set(LIBRARY_TYPE STATIC)
+if(BUILD_SHARED_LIBS)
+	set(LIBRARY_TYPE SHARED)
```
Comment (climbfuji) [inline src/CMakeLists.txt]:
```suggestion
  set(LIBRARY_TYPE SHARED)
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -215,23 +215,31 @@ if(((CMAKE_BUILD_TYPE STREQUAL "Debug") OR (CMAKE_BUILD_TYPE STREQUAL "RelWithDe
...
+endif()
+
+# Conditional flags based on the build type and compiler
+if(((CMAKE_BUILD_TYPE STREQUAL "Debug") OR (CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")) AND CMAKE_Fortran_COMPILER_ID STREQUAL "Intel")
+	SET_SOURCE_FILES_PROPERTIES(RTSolution/ADA/ADA_Module.f90 APPEND_STRING PROPERTY COMPILE_FLAGS " -check none -check bounds,pointers,uninit")
```
Comment (climbfuji) [inline src/CMakeLists.txt]:
```suggestion
  SET_SOURCE_FILES_PROPERTIES(RTSolution/ADA/ADA_Module.f90 APPEND_STRING PROPERTY COMPILE_FLAGS " -check none -check bounds,pointers,uninit")
```

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -215,23 +215,31 @@ if(((CMAKE_BUILD_TYPE STREQUAL "Debug") OR (CMAKE_BUILD_TYPE STREQUAL "RelWithDe
...
-
 # Fortran module output directory for build and install interfaces
 set(MODULE_DIR module/${PROJECT_NAME}/${CMAKE_Fortran_COMPILER_ID}/${CMAKE_Fortran_COMPILER_VERSION})
 set_target_properties(${PROJECT_NAME} PROPERTIES Fortran_MODULE_DIRECTORY ${CMAKE_BINARY_DIR}/${MODULE_DIR})
 install(DIRECTORY ${CMAKE_BINARY_DIR}/${MODULE_DIR}/ DESTINATION ${MODULE_DIR})
 target_include_directories(${PROJECT_NAME} INTERFACE
-                                $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/${MODULE_DIR}>
-                                $<INSTALL_INTERFACE:${MODULE_DIR}>)
-
+	$<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/${MODULE_DIR}>
```
Comment (climbfuji) [inline src/CMakeLists.txt]: Do you really want tabs here? Better to use whitespaces throughout.

**Review by climbfuji — COMMENTED** [review]
I can't really test this - someone from JCSDA should do that with jedi-bundle, and someone from EMC with their application. What I can do is try to add this to spack/spack-stack, configure the build variant shared accordingly, and then test building.

File: src/CMakeLists.txt (hunk 3 above)
Comment (BenjaminTJohnson) [reply src/CMakeLists.txt]: Agree completely, I need to find a way to default tab-to-whitespace in emacs.   I will fix it shortly.

**Comment by ADCollard** [comment]
I have so far build the CRTM with both options.  `src/libcrtm.a` and `src/libcrtm.so` are produced as appropriate.   `make install` still tries to put the libraries in `/usr/local/lib64`, which I do not have access to.

Maybe these instructions can be added to the README files?

I will now test in the GSI.

**Comment by BenjaminTJohnson** [comment]
@ADCollard I think we'll have to rely on passing `-DCMAKE_INSTALL_PREFIX=./` for local installs -- setting a default was causing issues with `spack-stack`.

**Comment by BenjaminTJohnson** [comment]
updated coeff tarball to `fix_REL-3.1.1.2` - this is the largest the tarball will ever be, and future tarballs will deprecate binary formats.   Almost all netCDF SpcCoeff and TauCoeff files are available.   see issue #101 for related discussion.

**Comment by BenjaminTJohnson** [comment]
converted to draft until I fix some lingering issues with "make install".

**Comment by BenjaminTJohnson** [comment]
@ADCollard @chengdang @emilyhcliu  This PR is ready for review.  @climbfuji  once this PR is in, this will be what we would want to use in spack stack.   I'm assuming it's too late for 1.8, but we can go for the next release.

**Comment by climbfuji** [comment]
You are lucky that we are behind with the release. I'll take a look and if it builds/installs fine, we can add it in

**Comment by climbfuji** [comment]
@BenjaminTJohnson Is there a new "fix" package (`crtm_fix`) that will go along with this version?

**Comment by BenjaminTJohnson** [comment]
@climbfuji  https://bin.ssec.wisc.edu/pub/s4/CRTM/fix_REL-3.1.1.2.tgz  and https://bin.ssec.wisc.edu/pub/s4/CRTM/fix_REL-3.1.1.2.tgz.md5sum 

Is that what you're looking for?

**Comment by climbfuji** [comment]
> @climbfuji https://bin.ssec.wisc.edu/pub/s4/CRTM/fix_REL-3.1.1.2.tgz and https://bin.ssec.wisc.edu/pub/s4/CRTM/fix_REL-3.1.1.2.tgz.md5sum
> 
> Is that what you're looking for?

Yes. Can you help me with the nitty gritty details for those fix file packages, please? For the various 2.4.0 versions, there are quite a few bug fixes/modifications in the `crtm_fix` package in spack:

https://github.com/JCSDA/spack/blob/d8de455f35c23ecbb2e53eacd661b7060145c122/var/spack/repos/builtin/packages/crtm-fix/package.py#L40

Thanks!

**Comment by climbfuji** [comment]
@BenjaminTJohnson I have several questions or issues with this version:
1. During the install, the cmake build downloads a gigantic test file (takes forever to download on my otherwise fast cable internet and to extract), but it seems to throw it away afterwards - the final installation is only 20M in size. If that data isn't needed (is that the data that EMC refers to as `crtm-fix` and that gets installed separately/as a dependency; see my message above) then it shouldn't be downloaded unless `crtm` understands spack's logic to run the tests after install, and the user requests running them.
   ```
   > du -hs  envs/ue-oneapi-2024.1.2/install/oneapi/2024.1.2/crtm-3.1.1-7ltgepx/
   20M     envs/ue-oneapi-2024.1.2/install/oneapi/2024.1.2/crtm-3.1.1-7ltgepx/
   ```
2. The install directory structure is not correct. See crtm@2.4.0.1 (the "EMC version", which has the correct setup):
   ```
   > ls -lart envs/ue-oneapi-2024.1.2/install/oneapi/2024.1.2/crtm-2.4.0.1-m2gpwti/
   total 36
   drwxr-sr-x   2 dom dom 12288 Aug 26 19:49 include
   drwxr-sr-x   5 dom dom  4096 Aug 26 19:49 .
   drwxr-sr-x   3 dom dom  4096 Aug 26 19:49 lib
   drwxr-sr-x   4 dom dom  4096 Aug 26 19:49 .spack
   drwxr-xr-x 244 dom dom 12288 Aug 26 20:41 ..
   ```
   Fortran modules (`*.mod`) are in `include`, the library `libcrtm.a` is in `lib`, and so is the `cmake` folder. Conversely:
   ```
   > ls -lart envs/ue-oneapi-2024.1.2/install/oneapi/2024.1.2/crtm-3.1.1-7ltgepx/
   total 10668
   drwxr-xr-x 244 dom dom    12288 Aug 26 20:41 ..
   drwxr-sr-x   3 dom dom     4096 Aug 26 21:08 module
   drwxr-sr-x   3 dom dom     4096 Aug 26 21:08 cmake
   drwxr-sr-x   5 dom dom     4096 Aug 26 21:08 .
   -rwxr-xr-x   1 dom dom 10891744 Aug 26 21:08 libcrtm.so
   drwxr-sr-x   4 dom dom     4096 Aug 26 21:08 .spack
   ```
   with the `*.mod files` in `envs/ue-oneapi-2024.1.2/install/oneapi/2024.1.2/crtm-3.1.1-7ltgepx/module/crtm/Intel/2021.0.0.20240211/` and the library/cmake files at the top-level directory. Is there a special prefix one has to pass to `cmake` to get the correct install tree? If so, why?
   
Thanks!

**Comment by BenjaminTJohnson** [comment]
> The install directory structure is not correct. See crtm@2.4.0.1 (the "EMC version", which has the correct setup)  

We haven't supported this structure since CRTM v2.x.  There won't be any special directory structure for EMC, and they're fine with the v3 structure.  Present testing by Andrew Collard / Emily Liu already conforms to the current stucture.

> cmake build downloads a gigantic test file

It downloads a 7 GB tarball of the coefficient files (`fix_REL-3.1.1.2.tgz`), and untars it.   The resulting files  end up in the `<build>/test_data/` directory.  This has been standard for all previous CRTMv3 releases.    The only difference is a 7GB tarball vs. a 5 GB tarball.  

> correct install tree

I guess this has made me realize that there was never a previous spack version of CRTMv3 (?)

I'm happy to make modifications (via CMake flag) to where the `.mod` files get installed.  

No changes to the CRTM directory structure will be made.   Everyone gets the same structure from now on as this vastly simplifies our support requirements.   GMAO and EMC have all adapted to the present v3 directory structure, and NRL (Hui C.) is on the path as well, although not sure how far along she is in her testing.

I'm not sure what to say about the coeffficient tarball download during the cmake step.  It's a large file, because we support a wide vareity of sensors, and (for this release only) 3 binary formats.    For users cloning from source, they only get this 1 time hit.    You should find a file named:   `fix_REL-3.1.1.2.tgz`  (that's what it's % counting during the cmake step.)   
All users also get the same tarball -- no separate EMC tarballs will be made.  If EMC has specific requirements for the content of the tarball, then we will add those to the coefficient tarball and update the 3.1.1.x version.     This needs to be communicated to me directly through EMC/NCO.  

For CRTM v3.2 (December), the size of that file should be about 2G, down from 7G.    We're in the binary to netCDF transition period, hence the temporary bloat.

If getting this particular release into spack-stack is too problematic, I don't have any problems with waiting until a future release.

**Comment by climbfuji** [comment]
> > The install directory structure is not correct. See crtm@2.4.0.1 (the "EMC version", which has the correct setup)
> 
> We haven't supported this structure since CRTM v2.x. There won't be any special directory structure for EMC, and they're fine with the v3 structure. Present testing by Andrew Collard / Emily Liu already conforms to the current stucture.
> 
> > cmake build downloads a gigantic test file
> 
> It downloads a 7 GB tarball of the coefficient files (`fix_REL-3.1.1.2.tgz`), and untars it. The resulting files end up in the `<build>/test_data/` directory. This has been standard for all previous CRTMv3 releases. The only difference is a 7GB tarball vs. a 5 GB tarball.
> 
> > correct install tree
> 
> I guess this has made me realize that there was never a previous spack version of CRTMv3 (?)
> 
> I'm happy to make modifications (via CMake flag) to where the `.mod` files get installed.
> 
> No changes to the CRTM directory structure will be made. Everyone gets the same structure from now on as this vastly simplifies our support requirements. GMAO and EMC have all adapted to the present v3 directory structure, and NRL (Hui C.) is on the path as well, although not sure how far along she is in her testing.
> 
> I'm not sure what to say about the coeffficient tarball download during the cmake step. It's a large file, because we support a wide vareity of sensors, and (for this release only) 3 binary formats. For users cloning from source, they only get this 1 time hit. You should find a file named: `fix_REL-3.1.1.2.tgz` (that's what it's % counting during the cmake step.) All users also get the same tarball -- no separate EMC tarballs will be made. If EMC has specific requirements for the content of the tarball, then we will add those to the coefficient tarball and update the 3.1.1.x version. This needs to be communicated to me directly through EMC/NCO.
> 
> For CRTM v3.2 (December), the size of that file should be about 2G, down from 7G. We're in the binary to netCDF transition period, hence the temporary bloat.
> 
> If getting this particular release into spack-stack is too problematic, I don't have any problems with waiting until a future release.

It almost sounds like we should spend a little time together figuring out how we can get crtm to install via spack so that it looks like any other cmake package (or any other package) with `lib`, `lib/cmake`, `include`, ... - we can make that an option in the cmake build, as long as future CRTMv3 versions will keep supporting that there's no problem. The directory structure shouldn't matter from a user's perspective, because the module file will take care of setting include paths, library paths, cmake paths etc. But it matters from a developer's point of view and we should install packages in spack following the industry standard as closely as possible.

Similar, for the tarball, it would make sense treating that as a separate package in spack `crtm-fix` as previous CMRT releases for EMC.

My suggestion is to spend a little time over the next weeks to figure that out and then do a post-install of CRTMv3 in the existing 1.8.0 release (e.g. as a chained environment).

**Comment by BenjaminTJohnson** [comment]
@climbfuji  That works for me.    I never had a separate repo/release for the fix files, I think this is something NCO was doing?    We only have one source for binary distribution, and that's the S4 FTP site.    

I'm generally available on Tuesdays/Thursdays.   Feel free to set a calendar invite that works for you.

**Comment by climbfuji** [comment]
> @climbfuji That works for me. I never had a separate repo/release for the fix files, I think this is something NCO was doing? We only have one source for binary distribution, and that's the S4 FTP site.
> 
> I'm generally available on Tuesdays/Thursdays. Feel free to set a calendar invite that works for you.

I'll do, thanks! After the spack-stack-1.8.0 release :-)

**Review by climbfuji — APPROVED** [review]
Approved based on my test install in spack-stack and the plan we laid out in the discusison below to address the few remaining issues that I came across.

**Comment by BenjaminRuston** [comment]
@climbfuji understand this came from an EMC request but may be good to discuss this ahead of time

I fully expect everyone including NOAA is going to be unloading the CRTM at some point and using their own, so we could discuss this *need* for it to be included and see if there's another way that satisfies

**Comment by climbfuji** [comment]
@BenjaminRuston I am not sure I understand. Are you trying to say that everybody is going to be expected to install crtm manually or as part of their modeling code instead of via spack-stack? That would be an unrealistic expectation I am afraid, and against the overall direction everyone is going. NOAA, NRL and other partners need a release of CRTM ready to load and use and certainly don't want to go through the hassle of installing it as part of the code that they are building.

**Review by chengdang — APPROVED** [review]
It is a great updates to the current CRTM build system.

One thing I noticed, with standalone mode (gfortran, MacOS): if CRTM is built with the debug mode (such as the given example` cmake -DCMAKE_BUILD_TYPE=DEBUG -DCMAKE_SHARED_LIBS=OFF -DCMAKE_INSTALL_PREFIX=./install ..`), the following tests fail, all due to errors in writing netCDF variables. It seems to be a netCDF/complier library/flag issue that I have not found a fix to it yet:

```
95% tests passed, 9 tests failed out of 177

Label Time Summary:
CRTM_Tests    = 527.97 sec*proc (177 tests)

Total Test time (real) = 112.79 sec

The following tests FAILED:
 4 - test_Unit_Aerosol_Bypass (Failed)
16 - test_forward_Simple_atms_n21 (Failed)
17 - test_forward_Simple_cris-fsr_n21 (Failed)
18 - test_forward_Simple_v.abi_g18 (Failed)
19 - test_forward_Simple_atms_npp (Failed)
20 - test_forward_Simple_cris399_npp (Failed)
21 - test_forward_Simple_v.abi_gr (Failed)
22 - test_forward_Simple_abi_g18 (Failed)
23 - test_forward_Simple_modis_aqua (Failed)
```

With the release flags, all tests passed.

How were these flags set? Is it based on some existing JEDI files?
https://github.com/JCSDA/CRTMv3/tree/feature/btj_REL-3.1.1_shared_vs_static_build/cmake

**Review by ADCollard — APPROVED** [review]
Approved.   

Was able to build and install both as fixed and shared libraries.    Was able to call from a standalone GSI job which completed successfully.

**Review by emilyhcliu — APPROVED** [review]
Same as @ADCollard 
Tested with stand-alone GSI with the updated coefficient packages 
Test completed successfully.

**Pull request merged**